[Justin Cheng] iP

Chustinjeng opened the pull request:
### JustinBot
> "Your mind is for having ideas, not holding them" - David Allen ([source](https://dansilvestre.com/productivity-quotes))

Remember when you were trying to remember what you had to do later, and you completely forgot the brilliant idea you once had? Yeah, me neither.

JustinBot solves this problem by ***freeing your mind*** of remembering what things you need to do. It's,
- text-based
- easy to learn
- ~FAST~ _SUPER_ FAST to use

All you need to do is,

1. Download it from [here](https://github.com/Chustinjeng/ip/releases/tag/A-Jar).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😀 

AND it is **FREE!**

Features:
- [x] Managing tasks
- [ ] Managing deadlines **(coming soon)**
- [ ] Reminders **(coming soon)**

If you are a Java programmer, you can use it to practise Java too! Here is the `main` method:
```java
public class Justin {
    public static void main(String[] args) {
        new JustinBot("data/justin.txt").run();
    }
}
```

Commit:
Add Gradle support

Chustinjeng committed:
no message

Chustinjeng committed:
Level-3 Level-4

Chustinjeng committed:
Level-5

Chustinjeng committed:
Level-6

Chustinjeng committed:
Level-1 Level-2

Chustinjeng committed:
no message

Chustinjeng committed:
Add FileWriter and method to write to files

Chustinjeng committed:
Added LocalDate formatting

Chustinjeng committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Added LocalDate formatting

# Conflicts:
#	src/main/java/Duke.java

Chustinjeng committed:
no message

Chustinjeng committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

* 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip:
  Add Gradle support

Chustinjeng committed:
Added JUnit tests

Chustinjeng committed:
Added new classes and packages

Chustinjeng committed:
Added Javadocs

Chustinjeng committed:
Corrected some coding mistakes

Chustinjeng committed:
Added JavaDocs

Chustinjeng committed:
Added Level-9: Find

Chustinjeng committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Added JavaDocs
  Added Javadocs

# Conflicts:
#	src/main/java/duke/Storage.java

Chustinjeng committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Added Level-9: Find

# Conflicts:
#	src/main/java/duke/Parser.java

Chustinjeng committed:
no message

Chustinjeng committed:
Made changes to Junit tests

Chustinjeng committed:
Added jar file

Chustinjeng committed:
Changed main class name to JustinBot and added new packages

e1010101 commented on src/main/java/justin/Parser.java:
```diff
@@ -0,0 +1,182 @@
+    }
+
+
+    public static Command parseFind(String message) {
```
Nit: missing Javadoc

e1010101 commented on src/main/java/JustinBot.java:
```diff
@@ -0,0 +1,64 @@
+     * parsing commands, which leads to actions
+     * being triggered in the program.
+     */
+    public void run() {
+        ui.welcome();
+        boolean isExit = false;
+        while (!isExit) {
+            try {
+                String fullCommand = ui.readCommand();
+                ui.showLine();
+                Command c = Parser.parse(fullCommand);
+                c.execute(tasks, ui, storage);
+                isExit = c.isExit();
+            } catch (DukeException e) {
+                ui.showText(e.toString());
+            } finally {
+                ui.showLine();
+            }
+        }
+    }
```
Clean initialisation process, nice!

e1010101 commented on src/main/java/justin/Parser.java:
```diff
@@ -0,0 +1,182 @@
+     * @throws DukeException if the description of the ToDo is empty.
+     */
+    public static Command parseToDo(String message) throws DukeException{
+        Scanner sc = new Scanner(message);
```
A global Scanner class can be used to avoid repetition.

sltsheryl replied:
Second this. Then use switch statements to return different types of tasks (which is like what you did in the parse(String input) method. I think there's a more efficient way to do this.

e1010101 commented on src/main/java/justin/Ui.java:
```diff
@@ -0,0 +1,153 @@
+
+    private final static String SEPARATOR = System.lineSeparator();
+
+    private final static String WELCOME_MESSAGE = "Hi, I'm Justin. What do you want me to do for you?";
+    private final static String GOODBYE_MESSAGE = "Bye! Hope to see you again soon.";
+    private final static String UNMARK_MESSAGE = "OK, I've marked your task as undone: ";
+    private final static String MARK_MESSAGE =  "Nice! I have marked this task as done: ";
+    private final static String LIST_MESSAGE = "Here are the tasks in your list: ";
+    private final static String DELETE_MESSAGE = "OK, I have removed the following task from the list: ";
+    private final static String ADD_MESSAGE = "Got it, I have added the following into the list: ";
+    private final static String FIND_MESSAGE = "Here are the matching tasks in your list: ";
```
Nit: could these messages possibly be moved to each corresponding command class?

sltsheryl replied:
Hmmm just a thought about this, would it be better to have all the messages in the Ui class since the class is supposed to deal with what's shown to the user? At the same time, I also feel putting each in their individual command class is neater and more direct?

tienyu2000 replied:
Proper declaration of constants.

e1010101 commented on src/main/java/justin/Ui.java:
```diff
@@ -0,0 +1,153 @@
+     * @param out A PrintStream to output messages.
+     */
+    public Ui(InputStream in, PrintStream out) {
+        this.in = new Scanner(in);
```
What is the rationale behind using a custom InputStream instead of System.in?

e1010101 commented on src/main/java/justin/command/FindCommand.java:
```diff
@@ -0,0 +1,18 @@
+    }
+
+    @Override
+    public void execute(TaskList list, Ui ui, Storage storage) {
+        ui.findMessage(list, description);
+    }
```
Should the find functionality be shifted here from the UI class?

e1010101 commented on src/main/java/justin/command/NewCommand.java:
```diff
@@ -0,0 +1,22 @@
+ */
+public class NewCommand extends Command {
+
+    /**
+     * Executes the command by doing nothing.
+     * @param list The TaskList to carry out operations.
+     * @param ui The Ui to send outputs.
+     * @param storage The Storage to save changes.
+     */
+    @Override
+    public void execute(TaskList list, Ui ui, Storage storage) {}
```
Perhaps more could be added to the Javadoc to clarify the purpose of this command?

e1010101 reviewed (commented):
Overall, very cohesive and clear implementation. Perhaps some refactoring is needed to ensure that classes cover the full scope of their responsibilities, but otherwise, really well done!

sltsheryl commented on src/main/java/JustinBot.java:
```diff
@@ -0,0 +1,64 @@
+
+    /**
+     * The main method to start the program.
+     * @param args Unused.
```
I think it's not specified in the coding standards but some places said it's not necessary to have a @param tag for main 
https://docs.oracle.com/javase/tutorial/getStarted/application/

sltsheryl commented on src/main/java/JustinBot.java:
```diff
@@ -0,0 +1,64 @@
+import justin.TaskList;
+import justin.Ui;
+
+/**
```
Well-documented and clear implementation within this class

sltsheryl commented on src/main/java/justin/Storage.java:
```diff
@@ -0,0 +1,96 @@
+     * @return the data file in String.
+     * @throws IOException if the file does not exist.
+     */
+    public String read() throws IOException {
```
The method name could be more suggestive such as readFile()

sltsheryl commented on src/main/java/justin/Storage.java:
```diff
@@ -0,0 +1,96 @@
+     * @throws DukeException if the file or directory does not
+     * exist.
+     */
+    public void save(TaskList list) throws DukeException {
```
Likewise, method naming could be more specific such as saveToFile(TaskList list)

sltsheryl commented on src/main/java/justin/Ui.java:
```diff
@@ -0,0 +1,153 @@
+
+/**
+ * Represents a class that is responsible for
+ * sending outputs to users.
```
I think could be more representative of Ui, as it sounds better to say it handles what is displayed on the user interface?

sltsheryl commented on src/test/java/justin/ToDoTest.java:
```diff
@@ -0,0 +1,20 @@
+
+public class ToDoTest {
+    @Test
+    public void dummyTest1() {
```
Could name your unit testing in a more suggestive way? Like valid/ invalid and should return what kind of output

sltsheryl commented on src/test/java/justin/DeadlineTest.java:
```diff
@@ -0,0 +1,22 @@
+public class DeadlineTest {
+
+    @Test
+    public void dummyTest1() {
```
Comprehensive testing, but you could give them better naming :)

sltsheryl commented on src/main/java/justin/command/DeleteCommand.java:
```diff
@@ -0,0 +1,43 @@
+    private int num;
+
+    /**
+     * Constructor for the DeleteCommand class.
```
I checked. Your command classes are documented really comprehensively 👍🏻 for overridden methods, prof said it's not necessary but up to us.

sltsheryl commented on src/main/java/justin/command/Command.java:
```diff
@@ -0,0 +1,31 @@
+     * Returns the boolean value isExit in the Command object.
+     * @return boolean isExit.
+     */
+    public boolean isExit() {
```
Good naming to show it's a boolean

sltsheryl reviewed (commented):
You did a really good job implementing and documenting your code. Most of it is well-written just here and there, you could improve your naming. And could think of alternative implementation such as the parse to make it shorter/ less repetitive, additional work done!

kevinchangjk commented on src/main/java/JustinBot.java:
```diff
@@ -0,0 +1,64 @@
+
+    /**
+     * Constructor for the Duke class.
+     * @param filePath The path of the data file
```
I think an empty line between the description and the parameter sections should be better? Might help with readability, especially since your descriptions tend to be "wrapped" in a few lines. This applies for other javadoc comments too.

kevinchangjk commented on src/main/java/justin/Parser.java:
```diff
@@ -0,0 +1,182 @@
+        Scanner sc = new Scanner(message);
+        String first = sc.next();
+        switch (first) {
+            case "bye": {
```
For our coding standards, I think we're having the `case` clauses on the same indent level as the `switch` statements. So maybe something like this?

```java
        // ...
        switch (first) {
        case "bye": {
            return new ExitCommand();
        }
        // ...
```

If you are using an IDE, I think it should be configurable somewhere in the settings.

tienyu2000 replied:
Second this, can right click and select "column selection mode" in Intellij to position case clause at the same indent level as switch statements.

kevinchangjk commented on src/main/java/justin/Parser.java:
```diff
@@ -0,0 +1,182 @@
+                lt = LocalTime.parse(arr[3], formatter);
+                return new AddEventCommand(description, isDone, ld.toString(), lt.toString());
+            default:
+                throw new DukeException("OOPS, looks like the file has been corrupted. Please delete the file and try again.");
```
Minor nit, this line seems to be like 6 chars too long, maybe can instantiate the error message separately on another line?

kevinchangjk commented on src/main/java/justin/Storage.java:
```diff
@@ -0,0 +1,96 @@
+            System.out.println("File is corrupted. Please delete tasks file and try again.");
+            System.exit(0);
+        }
+        String[] strArray = tasksStr.split(System.lineSeparator());
```
Hmm I kinda get that `strArray` literally means an array, but maybe the naming could more explicitly hint a collection of items? Something in the plural form?

kevinchangjk commented on src/main/java/justin/TaskList.java:
```diff
@@ -0,0 +1,101 @@
+            curr.mark();
+        } catch (ArrayIndexOutOfBoundsException e) {
+            throw new DukeException("OOPS! Please enter a valid number!");
+        } catch (Exception e) {
```
From the error message, seems like you're anticipating a missing param exception, so maybe the catch statement can reflect that, instead of catching the generic `Exception` class?

kevinchangjk commented on src/main/java/justin/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,68 @@
+public class AddDeadlineCommand extends Command {
+    private String description;
+    private boolean isDone;
+    private String by;
```
Perhaps the naming for the `by` variable can be more consistent with that of the `time` variable? I notice that both refer to time (technically one date and one time), so maybe for readability they should have similar names, because "by" doesn't really seem to relate to "time".

tienyu2000 replied:
Good naming to show variable is a boolean.

kevinchangjk commented on src/main/java/justin/task/Task.java:
```diff
@@ -0,0 +1,52 @@
+     * @return String representation of Task's completion.
+     */
+    public String getStatusIcon() {
+        return (isDone ? "X" : "O");
```
Minor nit, but `isDone` here is referred to as `this.isDone` in other places, so maybe this could be more consistent.

kevinchangjk reviewed (commented):
Very nice work, and like the others mentioned, really good documentation. Just a few nits here and there, but otherwise LGTM.

Chustinjeng committed:
Added javafx GUI

Chustinjeng committed:
Added javadocs

Chustinjeng committed:
Added Varargs

Chustinjeng committed:
Add assertion statements to various methods to signify important
assumptions.

Some important assumptions are not stated clearly in code and cannot be
identified by users easily.

Let's place some assertion statements to improve code readability and
allow users to understand some important assumptions.

Chustinjeng committed:
Improve code quality and readability through better Single Level of
Abstraction principle.

There are some methods that did not utilise the Single Level of
Abstraction principle as they provide multiple levels of abstraction.

Readability of code is not good if there are multiple levels of
abstraction.

Let's improve abstraction levels by creating more methods to encapsulate
their meaning in code.

Chustinjeng committed:
Merge pull request #1 from Chustinjeng/branch-A-CodeQuality

Improve code quality and readability through better Single Level of Abstraction

Chustinjeng committed:
Merge branch 'master' into branch-A-Assertions

Chustinjeng committed:
Merge pull request #2 from Chustinjeng/branch-A-Assertions

Add assertion statements to various methods to signify important assumptions

Chustinjeng committed:
Added the ability to identify overlaps in events

Chustinjeng committed:
Made final edits to GUI and the bot.

Chustinjeng committed:
no message
